### PR TITLE
fix: improve transitive dependency error messages and skip primitives in auto-injecting factory

### LINKED ISSUE #489: Misleading AdapterNotFoundError when transitive dependency fails
## Bug

When an adapter IS found for a type but its constructor dependency fails to resolve, `_rust_core.resolve()` raises a `KeyError` that gets wrapped as `AdapterNotFoundError` for the *top-level* type, not the missing dependency. This sends the developer to the wrong place.

## Repro

```python
# settings_port.py
class SettingsPort(Protocol):
    some_field: str

# auth.py
@adapter.for_(AuthPort, profile=Profile.DEVELOPMENT)
class ClerkAdapter:
    def __init__(self, settings: SettingsPort) -> None: ...

# settings.py
@adapter.for_(SettingsPort, profile=Profile.DEVELOPMENT)
class EnvSettings(pydantic.BaseSettings):
    some_field: str = "default"
```

When `AuthPort` is resolved with `Profile.DEVELOPMENT`, `ClerkAdapter` is found but its transitive dependency `SettingsPort` fails to resolve (e.g., the `EnvSettings` pydantic adapter can't be constructed). The error is:

```
AdapterNotFoundError: No adapter for AuthPort in profile 'development'
  Registered: ClerkAdapter (development, production), FakeAuthAdapter (test)
```

This says "No adapter for AuthPort" — but `ClerkAdapter` IS registered for `development`. The real error is chained inside as a `KeyError` about the transitive resolution failure.

## Expected

The error should clearly identify the resolution chain that failed:

```
AdapterNotFoundError: Cannot resolve AuthPort → ClerkAdapter → settings: SettingsPort
  ClerkAdapter found for AuthPort (development, production)
  SettingsPort resolution failed: <inner error>
```

At minimum, the outer error message should not claim there's "no adapter" for a type when one is clearly registered.

## Impact

This cost nontrivial debugging time. The developer sees "ClerkAdapter (development)" right there in the error message, dismisses AuthPort resolution as the problem, and digs into why dioxide "can't find" an adapter that's clearly listed.

### LINKED ISSUE #490: Can't construct pydantic BaseSettings adapters for transitive injection
## Bug

Adapters that inherit from `pydantic.BaseSettings` (or any class with `**kwargs` in `__init__`) fail to be auto-constructed when resolved as a transitive dependency. They work fine when explicitly pre-registered via `Container.register_instance()`.

## Repro

```python
@adapter.for_(SettingsPort, profile=[Profile.PRODUCTION, Profile.DEVELOPMENT])
class EnvSettings(BaseSettings):
    database_url: str = "postgresql://..."
    
    model_config = {"env_file": ".env"}

@adapter.for_(AuthPort, profile=[Profile.PRODUCTION, Profile.DEVELOPMENT])
class ClerkAdapter:
    def __init__(self, settings: SettingsPort) -> None:
        self._key = settings.clerk_secret_key
```

Resolving `AuthPort` → `ClerkAdapter` → needs `SettingsPort` → finds `EnvSettings`.

Dioxide finds `EnvSettings` as an adapter for `SettingsPort`, but fails to construct it. The inner error is:

```
KeyError: "Python error: TypeError: ClerkAdapter.__init__() missing 1 required positional argument: 'settings'"
```

## Root cause theory

`BaseSettings.__init__` accepts `**kwargs`. Dioxide's constructor analysis likely sees this and either:
- Can't determine what to pass, so it fails
- Tries to treat all pydantic model fields as injectable dependencies

Since pydantic models take zero *required* args (everything is `**kwargs`), the fallback of "call with no args" should work here.

## Workaround

```python
di_container = Container()
di_container.register_instance(SettingsPort, EnvSettings())
```

Pre-constructing and registering the settings instance works because it bypasses constructor resolution. But this shouldn't be necessary — if an adapter's constructor accepts `**kwargs` (meaning all args are optional), calling with no args should be a valid fallback.

## Suggestion

When constructor analysis encounters `**kwargs` with no required positional args, try zero-arg construction as a fallback before failing.

### PR DESCRIPTION
**Pull request by mikelane**
## Summary
- Fixes misleading `AdapterNotFoundError` when transitive dependency resolution fails — the error now shows the full resolution chain (e.g., `AuthPort → ClerkAdapter → settings: SettingsPort`) instead of falsely claiming no adapter exists
- Skips Python primitive types (str, int, float, bool, etc.), `Optional[primitive]` union types, and `**kwargs`/`*args` parameters in the auto-injecting factory, fixing auto-construction of pydantic `BaseSettings` adapters as transitive dependencies
- Adds recursion guard to prevent infinite loops on circular dependency resolution

## Changes
- `python/dioxide/container.py` — Added `_PRIMITIVE_TYPES` constant, `_is_primitive_or_optional_primitive`, `_is_registered_in_container`, `_build_transitive_failure_message`, `_resolving` guard, `_probing_deps` guard; modified `resolve()` and `_create_auto_injecting_factory`
- `tests/test_descriptive_errors.py` — 4 new tests for transitive dependency chain errors
- `tests/test_config_injection.py` — 5 new tests for primitive/optional/union handling
- `tests/test_adversarial_primitives.py` — 13 new adversarial QA tests
- `tests/test_terse_errors.py` — Updated line-count assertion

Closes #489
Closes #490

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: tests/test_gauntlet_adversarial_qa.py [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+state management, and edge cases of the primitive-type detection and
+transitive-failure probes.
+
+Each bug is proven by a failing test before recording.
+"""
+
+from __future__ import annotations
+
+from typing import (
+    Literal,
```
Comment (mikelane) [inline tests/test_gauntlet_adversarial_qa.py]: CRITICAL: Unused import Literal causes ruff lint failure (F401). The import at line 13 is never used in actual code — only referenced in comments and docstrings. Remove the Literal import from the typing imports block.

File: tests/test_gauntlet_adversarial_qa.py [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+        class MixedService:
+            def __init__(self, value: str | SettingsPort = 'default') -> None:
+                self.value = value
+
+        container = Container()
+        container.scan(profile=Profile.DEVELOPMENT)
+
+        # BUG: This crashes with TypeError instead of raising
+        # ServiceNotFoundError with a clear message.
+        with pytest.raises(TypeError, match="'Union' object is not an instance of 'type'"):
```
Comment (mikelane) [inline tests/test_gauntlet_adversarial_qa.py:59]: HIGH: This test asserts a KNOWN BUG — TypeError crash for Union[str | SettingsPort]. The test passes CI by asserting broken behavior. It should either be fixed in this PR, marked with @pytest.mark.xfail, or tracked as a separate issue. Shipping a test that validates incorrect behavior inflates coverage numbers and creates confusion about what was actually fixed.

The same issue applies to DescribeBugLiteralTypeCrash (line 93) and DescribeBugRequiredPrimitiveMisleadingMessage (line 131).

File: python/dioxide/container.py [hunk 3]
```diff
@@ -1158,6 +1225,186 @@ def _check_captive_dependencies(self, port_to_adapters: dict[type[Any], list[typ
...
+            # Guard against circular dependency probing
+            if dep_type in self._probing_deps:
+                failed_param = param_name
+                failed_type_name = getattr(dep_type, '__name__', str(dep_type))
+                failed_reason = 'circular dependency detected during resolution'
+                break
+
+            self._probing_deps.add(dep_type)
+            try:
+                self.resolve(dep_type)
```
Comment (mikelane) [inline python/dioxide/container.py:1373]: MEDIUM: Probing each dependency by calling self.resolve() creates side effects during error message construction. If a non-primitive dependency resolves successfully (not the failing one), it will be instantiated and potentially cached as a singleton. While this is functionally benign (the same instance would be created on a successful resolution), it is surprising behavior for an error-reporting method. Consider documenting this in the method contract or using a lighter-weight probe (e.g., checking registration without full instantiation).

File: python/dioxide/container.py [hunk 4]
```diff
@@ -1158,6 +1225,186 @@ def _check_captive_dependencies(self, port_to_adapters: dict[type[Any], list[typ
...
+            if '<locals>' in implementing_class.__qualname__:
+                import sys
+                from types import FrameType
+
+                frame: FrameType | None = sys._getframe()
+                while frame is not None:
+                    for name, obj in frame.f_locals.items():
+                        if inspect.isclass(obj):
+                            localns[name] = obj
+                    frame = frame.f_back
```
Comment (mikelane) [inline python/dioxide/container.py:1338]: LOW: This frame-walking pattern for resolving local class namespaces is identical to code in _create_auto_injecting_factory (lines 2890-2909). Consider extracting a shared helper (e.g., _get_local_class_namespace) to avoid the duplication. Both copies use the same sys._getframe() + inspect.isclass walk.

**Review by mikelane — COMMENTED** [review]
## PR Review: #491 -- Transitive Dependency Error Messages + Primitive Handling

### Summary
The PR correctly addresses the core bugs (#489 and #490): transitive dependency failures now show resolution chains, and auto-injecting factories skip primitive/optional-primitive parameters. The implementation is sound, the logic for distinguishing genuine missing types from transitive failures via _is_registered_in_container / _build_transitive_failure_message is well-structured, and the recursion guards (_resolving, _probing_deps) are correctly placed with proper cleanup.

However, there is one blocker and two significant issues that must be addressed before merge.

### Blocker
- **CRITICAL: Unused import Literal in test_gauntlet_adversarial_qa.py:13** -- ruff reports F401. This is a lint failure that must be fixed. Remove the import or annotate it with noqa if it serves a documentation purpose.

### High
- **Adversarial QA tests assert bugs, not fixes** -- test_gauntlet_adversarial_qa.py contains three test classes (DescribeBugUnionPrimitiveNonPrimitiveCrash, DescribeBugLiteralTypeCrash, DescribeBugRequiredPrimitiveMisleadingMessage) that assert current broken behavior (TypeError crashes and misleading messages). These tests pass by validating that bugs still exist. They inflate test counts (~800 lines) while verifying nothing was fixed. Either fix these bugs, mark them xfail, or move them to a separate tracking issue.

### Medium
- **_build_transitive_failure_message probing creates side effects** -- calling self.resolve() for each non-primitive dependency instantiates (and potentially caches as singletons) any dependency that resolves successfully before finding the failing one. Functionally benign but surprising for an error-reporting method.

### Low
- **Duplicate frame-walking code** -- _build_transitive_failure_message (lines 1329-1338) and _create_auto_injecting_factory (lines 2890-2909) share identical logic for resolving local class namespaces. Consider extracting a shared helper.
- **Relaxed line count assertion** -- test_terse_errors.py relaxed from 4 to 6 lines. Acceptable given the richer chain format.

### Praise
- The _is_registered_in_container / _build_transitive_failure_message separation correctly distinguishes genuinely missing types from registered types whose factories fail -- this is the right architectural approach.
- State cleanup in _resolving and _probing_deps is comprehensive -- the adversarial tests for state leakage (DescribeResolvingStateCleanup, DescribeProbingDepsCleanup) are well-designed.
- The Optional[primitive] handling (UnionType, Optional[X], Union[X, None]) is thorough and correctly uses typing.get_origin/get_args.
- Test coverage is extensive -- 88 new/modified tests covering the core fixes plus adversarial edge cases.

### Decision: REQUEST_CHANGES
Address the unused import and resolve the adversarial QA bug-assertion tests before re-review.

File: python/dioxide/container.py (hunk 3 above)
Comment (mikelane) [inline python/dioxide/container.py:1373]: **MEDIUM**: Probing dependencies by calling `self.resolve(dep_type)` at line 1373 has observable side effects.  

If the first dependency succeeds but a subsequent one fails, the successful dependency's singleton is created and cached in the Rust container — even though the overall resolve fails and the caller will retry. This is technically observable (the singleton is "wasted" in the cache) though it doesn't cause correctness issues since the resolution fails anyway.

The probing guard (`_probing_deps`) correctly prevents infinite recursion, but the `resolve()` call itself can mutate container state. Consider whether an explicit dry-run or `resolver.try_instantiate()` that doesn't cache singletons would be cleaner, or document this as intentional "probe-and-cleanup" behavior.

File: python/dioxide/container.py [hunk 5]
```diff
@@ -1158,6 +1225,186 @@ def _check_captive_dependencies(self, port_to_adapters: dict[type[Any], list[typ
...
+        failed_reason: str | None = None
+
+        try:
+            init_signature = inspect.signature(implementing_class.__init__)
+            globalns = getattr(implementing_class.__init__, '__globals__', {})
+            localns = dict(vars(implementing_class))
+            localns[implementing_class.__name__] = implementing_class
+
+            # Handle local classes in tests
+            if '<locals>' in implementing_class.__qualname__:
```
Comment (mikelane) [inline python/dioxide/container.py:1329]: **LOW**: This frame-walking block (lines 1329-1338) is a duplicate of the identical logic in `_create_auto_injecting_factory` (lines 2890-2909). Both walk the call stack looking for local class definitions to populate `localns` for `get_type_hints`.

Consider extracting into `_collect_frame_locals(qualname: str) -> dict[str, Any]` to eliminate the duplication. The two call sites differ only in the `localns` variable they start with.

File: python/dioxide/container.py [hunk 6]
```diff
@@ -933,37 +984,53 @@ def resolve(self, component_type: type[T]) -> T:
                 component_name = component_type.__name__
                 raise ScopeError(f'Cannot resolve {component_name}: REQUEST-scoped, requires active scope')
 
+        # Circular dependency guard: if this type is already in the resolve
+        # call stack, raise immediately to prevent infinite recursion.
+        if component_type in self._resolving:
```
Comment (mikelane) [inline python/dioxide/container.py:989]: **LOW**: The raw `KeyError` message text from the circular dependency guard (line 989: `'Circular dependency detected...'`) can leak through into the transitive failure chain message when the error is caught by the `except (AdapterNotFoundError, ServiceNotFoundError, KeyError)` clause at line 1374.  

The `failed_reason` field would show the raw `KeyError` string like `\"Circular dependency detected: 'ServiceX' is already being resolved\"`. While functional, a dedicated exception (or a structured error type) would avoid exposing raw Python exception text in user-facing messages.

File: tests/test_gauntlet_adversarial_qa.py [hunk 7]
```diff
@@ -0,0 +1,798 @@
...
+class DescribeBugUnionPrimitiveNonPrimitiveCrash:
+    """When a service constructor has a Union type that mixes a primitive
+    with a non-primitive (e.g. str | SomePort), _is_primitive_or_optional_primitive
+    returns False, the auto-injecting factory tries to resolve the entire Union
+    from the container, and _rust_core.resolve crashes with:
+    TypeError: argument 'py_type': 'Union' object is not an instance of 'type'
+
+    Expected: ServiceNotFoundError with a clear message."""
+
+    def it_crashes_with_typeerror_for_str_or_port_union(self) -> None:
```
Comment (mikelane) [inline tests/test_gauntlet_adversarial_qa.py:43]: **MEDIUM** (documented bug, not a regression): These adversarial QA tests correctly identify that Union types mixing primitives and non-primitives (e.g., `str | SettingsPort`) crash with TypeError instead of raising a clean `ServiceNotFoundError`.  

This is a legitimate gap in the #490 fix: `_is_primitive_or_optional_primitive` correctly handles `str | None` but not `str | SettingsPort`. A follow-up issue should track adding a validation step to ensure the auto-injecting factory either skips such params (using their default) or raises a descriptive error before passing Union special forms to the Rust core.

These tests should NOT block this merge — they're adversarial QA doing its job correctly: finding edge cases the implementation missed. They should be filed as follow-up issues.

File: tests/test_gauntlet_adversarial_qa.py [hunk 8]
```diff
@@ -0,0 +1,798 @@
...
+        error_msg = str(exc_info.value)
+        assert 'AnyPort' in error_msg
+
+
+# ============================================================
+# Attack 11: Deep transitive chains (depth >= 3)
+# ============================================================
+
+
+class DescribeDeepTransitiveChains:
```
Comment (mikelane) [inline tests/test_gauntlet_adversarial_qa.py:534]: **LOW** (naming): The class `DescribeDeepTransitiveChains` also appears in `tests/test_adversarial_primitives.py` at line 182. Both test files are in the same test suite and would be collected together. While pytest won't error (module paths disambiguate), identical class names make stack traces and test filtering ambiguous.  

Consider renaming one — e.g., `DescribeDeepTransitiveChainsQA` here or `DescribeDeepTransitiveChainDetection` in the other file.

File: python/dioxide/container.py (hunk 6 above)
Comment (mikelane) [inline python/dioxide/container.py:989]: **PRAISE**: The dual-guard pattern using `_resolving` at line 989 and `_probing_deps` at line 1365 is clean and correct. Two separate sets with distinct lifetimes serve different purposes: `_resolving` catches cycles in the `resolve()` call stack (prevents infinite factory recursion), while `_probing_deps` catches cycles specific to the error-probing walk (prevents infinite probe recursion). Both are cleaned up in `finally` blocks, ensuring state isolation between separate resolution attempts. This is exactly the right level of defense in depth.

**Review by mikelane — COMMENTED** [review]
## Review Verdict: APPROVE

**0 CRITICAL, 0 HIGH, 3 MEDIUM, 3 LOW** — no blockers. All quality tools pass (ruff, mypy, format check). All 902 tests pass with zero failures.

### What was reviewed

PR #491 at commit `7ecce0b` fixes two bugs: #489 (resolution chain in transitive dependency errors) and #490 (skip Python primitives in auto-injecting factories).

### Changes from previous review (de2fe61)

The CRITICAL `Literal` import issue reported in the previous review against commit `de2fe61` is FIXED. The Literal tests now use an external fixture module (`tests/fixtures/qa_literal_fixtures.py`) so `get_type_hints` can properly resolve the annotations, and the test correctly confirms the crash rather than masking it.

### Core fix assessment

The fix logic is correct and clean:

1. **`_is_primitive_or_optional_primitive`** — Handles bare primitives, `UnionType` (`str | None`), and `typing.Union` (`Optional[str]`). Also handles multi-primitive unions like `str | int | float`. Correctly returns False for `list[str]` generic aliases (which should be resolved from the container).

2. **`_build_transitive_failure_message`** — Probes each dependency individually to identify the specific failing one, then builds a chain-aware error message. The `_probing_deps` guard prevents infinite recursion during probing.

3. **Circular dependency guard** — The `_resolving` set in `resolve()` catches cycles early and raises a clear `KeyError` rather than allowing infinite recursion through factory closures.

4. **Auto-injecting factory** — Correctly skips both primitive-typed params and `VAR_KEYWORD`/`VAR_POSITIONAL` params. The `injectable_params` list is used both for the early-return optimization (return `cls` directly when no params need injection) and for the factory closure itself.

### Adversarial QA tests

The adversarial QA tests in `test_gauntlet_adversarial_qa.py` (Attack 1-3) correctly document three remaining edge cases that the implementation didn't handle:
- Union[primitive | non-primitive] crashes with TypeError
- typing.Literal types crash with TypeError
- Required primitives without defaults get a misleading error message

These are **not regressions** — the adversarial QA phase is designed to find bugs the implementation missed. These tests should NOT block merge but should have follow-up issues filed.

### Recommendations (non-blocking)

1. File follow-up issue for Union[primitive|non-primitive] TypeError crash (Attack 1)
2. File follow-up issue for typing.Literal TypeError crash (Attack 2)
3. Extract duplicate frame-walking logic into a shared helper method
4. Rename duplicate `DescribeDeepTransitiveChains` class in one of the two test files

### What's excellent

- The `_resolving`/`_probing_deps` dual-guard pattern for circular dependency safety
- Clean `finally` block cleanup ensuring state isolation between resolution attempts
- The `_is_primitive_or_optional_primitive` static method handles all Union variants correctly
- Comprehensive test coverage: unit tests (descriptive errors, config injection, terse errors), adversarial QA, and state cleanliness tests
- Documentation updated with chain-aware error examples
